### PR TITLE
CC | image pulling in the guest without forked containerd

### LINKED ISSUE #7790: CC | image pulling in the guest without forked containerd
**Which feature do you think can be improved?**

 image pull in the guest

**How can it be improved?**

image pulling in the guest without forked containerd

### PR DESCRIPTION
**Pull request by ChengyuZhu6**
This PR is to achieve `image pulling in the guest without forked containerd`. 
Without forked containerd, the image name will be passed from remote snapshotter to kata-agent through mount slice.

The following are operations and results with image pulling in the guest:
 
- snapshotter:  pass the image url to kata through mount options with prefix `io.kata-containers.volume=`:
```
"io.kata-containers.volume=eyJ2b2x1bWVfdHlwZSI6ImltYWdlX2d1ZXN0X3B1bGwiLCJzb3VyY2UiOiJxdWF5LmlvL2thdGEtY29udGFpbmVycy9jb25maWRlbnRpYWwtY29udGFpbmVyczp1bnNpZ25lZCIsImZzX3R5cGUiOiJvdmVybGF5ZnMiLCJpbWFnZV9wdWxsIjp7Im1ldGFkYXRhIjp7fX19"
``` 
the decoded json: 
```
{"volume_type":"image_guest_pull","source":"quay.io/kata-containers/confidential-containers:unsigned","fs_type":"overlayfs","image_pull":{"metadata":{}}}
```
- kata-runtime:  
1) set the rootfs storage: 
```
Storage { driver: "image_guest_pull", driver_options: ["image_guest_pull={"metadata":{}}"], source: "quay.io/kata-containers/confidential-containers:unsigned", fstype: "overlay", options: [], mount_point: "/run/kata-containers/f83078b04f585e6dd7e2964ece8b776ed7e1eef81ec76981e90fbd3258f01614/rootfs"...
```
- kata-agent:  
1) call `image-rs` to pull image to a bundle path: `/run/kata-containers/f83078b04f585e6dd7e2964ece8b776ed7e1eef81ec76981e90fbd3258f01614/images/rootfs`. 
2) bind mount the bundle path to the mountpoint `/run/kata-containers/f83078b04f585e6dd7e2964ece8b776ed7e1eef81ec76981e90fbd3258f01614/rootfs` by `overlayfs`.

Notes: The reason for redefining functions of `image_rpc` and the distinction from `PullImageRequest` stems from the initialization of `rootfs.storage` during the kata-runtime stage. In the context of kata-agent, failing to redefine the functions of `image_rpc` can result in a crash of the kata-agent when removing containers. This occurs because, during the removal of container resources, kata-agent attempts to unmount the rootfs mountpoint located at `/run/kata-containers/cid/rootfs`. While the `image_rpc` operation `pull_image `is primarily concerned with pulling and unpacking images to the specified mountpoint. To address this, a modification is made to the bundle path, relocating it to `/run/kata-containers/cid/images/rootfs`. Subsequently, we bind mount the path to `/run/kata-containers/cid/rootfs`. This adjustment ensures the proper handling of container removal, preventing crashes within the kata-agent and facilitating a smooth cleanup process.
 
depends: https://github.com/containerd/nydus-snapshotter/pull/499

Fixes #7790

### CODE REVIEW

File: src/agent/src/image_rpc.rs
Comment (stevenhorsman) [inline src/agent/src/image_rpc.rs:284]: Sorry to be a pain, but as this code is moved from earlier in the file I wondered if it would be possible to preserve the authorship and sign-off of the moved lined by adding them to this commit as I think that's good practice? Equally if anyone thinks this is wrong/excessive, please let me know!
I took a look at the moved lines and the full list of people involved in them is:
```
Co-authored-by: Jiang Liu <gerry@linux.alibaba.com>
Co-authored-by: Wang, Arron <arron.wang@intel.com>
Co-authored-by: wllenyj <wllenyj@linux.alibaba.com>
Co-authored-by: jordan9500 <jordan.jackson@ibm.com>
Co-authored-by: stevenhorsman <steven@uk.ibm.com>
```

File: src/runtime/virtcontainers/fs_share_linux.go
Comment (stevenhorsman) [inline src/runtime/virtcontainers/fs_share_linux.go:447]: Can I ask for a bit of clarity here. Is the way this is working that `/run/kata-containers/` from the host is mounted to the guest and then used for pull the image too, or something else?

**Review by stevenhorsman — COMMENTED** [review]
Sorry, I'm probably just being slow and not understanding, or missing something, but It would be great to have a bit for information in the 2nd and 3rds commits to explain a bit more about what the code does and why to help me out with understanding it and reviewing. The image_rpcs changes that I understand more look okay me to though. Thanks!

File: src/runtime/virtcontainers/fs_share_linux.go
Comment (ChengyuZhu6) [reply src/runtime/virtcontainers/fs_share_linux.go:447]: @stevenhorsman  No, that simply sets the rootfs mountpoint for the container within the guest, which will subsequently perform a bind mount of the directory fetched by the image_rpc to this mountpoint.

File: src/agent/src/image_rpc.rs
Comment (ChengyuZhu6) [reply src/agent/src/image_rpc.rs:284]: > Done.

**Comment by ChengyuZhu6** [comment]
> Sorry, I'm probably just being slow and not understanding, or missing something, but It would be great to have a bit for information in the 2nd and 3rds commits to explain a bit more about what the code does and why to help me out with understanding it and reviewing. The image_rpcs changes that I understand more look okay me to though. Thanks!

Thanks for review. I'm sorry for that. I have add more information in the commits.

**Comment by ChengyuZhu6** [comment]
`
jenkins-ci-ubuntu-20-04` error:
```
Build kata version 3.2.0-rc0: shim-v2
INFO: DESTDIR /tmp/jenkins/workspace/kata-containers-CCv0-ubuntu-20.04-PR/go/src/github.com/kata-containers/kata-containers/tools/packaging/kata-deploy/local-build/build/shim-v2/destdir
Error response from daemon: manifest for quay.io/kata-containers/cc-builders:shim-v2-go-1.19.3-rust-1.69.0-5f5e05a77f5122db3499d13861ad2b34a61e0ee9-x86_64 not found: manifest unknown: manifest unknown
```

File: src/agent/src/storage/image_pull_handler.rs
Comment (stevenhorsman) [inline src/agent/src/storage/image_pull_handler.rs:1]: Thanks for the extra information in the commits, it was very helpful for me to understand what is going on, but just to double check that I've got it correctly - this is what I think it happening in this commit:

With the nydus/other snapshotter approach, if we are using image offload then we will receive the image metadata in createContainerRequest in rootfs.storage under the `KATA_VIRTUAL_VOLUME_IMAGE_GUEST_PULL` key (which maps to `image_guest_pull`?) as a json object and the image name in storage.source? This all set up by a combination of the snapshotter passing through via extraoptions and then the runtime converting it into this format in create.go in the next commit?

What’s the mechanism that means this agent logic only runs if `driver: "image_guest_pull”`?

File: src/agent/src/storage/image_pull_handler.rs
Comment (stevenhorsman) [inline src/agent/src/storage/image_pull_handler.rs:23]: It might be helpful to add a unit test for the `get_image_info` to show an example of the expected payload in close proximity to the code (and obviously to test the capability).

**Review by stevenhorsman — COMMENTED** [review]
Thanks for the extra commentary, it was really helpful. I just have a couple of extra clarifying questions/comments

File: src/agent/src/storage/image_pull_handler.rs
Comment (ChengyuZhu6) [reply src/agent/src/storage/image_pull_handler.rs:23]: Sure.

File: src/agent/src/storage/image_pull_handler.rs
Comment (ChengyuZhu6) [reply src/agent/src/storage/image_pull_handler.rs:1]: Hi, @stevenhorsman . I have updated my commits.  The first eight commits are same as https://github.com/kata-containers/kata-containers/pull/7688. The last three commits implement the feature of pulling image in the guest. Please help me to review again.

Comment (ChengyuZhu6) [reply src/agent/src/storage/image_pull_handler.rs:1]: >What’s the mechanism that means this agent logic only runs if driver: "image_guest_pull”?
we create a new storage handlers in agent to handle `image_guest_pull`:
```
pub static ref STORAGE_HANDLERS: StorageHandlerManager<Arc<dyn StorageHandler>> = {
......
        manager.add_handler(KATA_VOLUME_TYPE_DMVERITY, Arc::new(DmVerityHandler{})).unwrap();
        manager.add_handler(KATA_VIRTUAL_VOLUME_IMAGE_GUEST_PULL, Arc::new(ImagePullHandler{})).unwrap();
        manager
    };
```

Comment (ChengyuZhu6) [reply src/agent/src/storage/image_pull_handler.rs:1]: >(which maps to image_guest_pull?)  
yes.

Comment (ChengyuZhu6) [reply src/agent/src/storage/image_pull_handler.rs:1]: >With the nydus/other snapshotter approach, if we are using image offload then we will receive the image metadata in createContainerRequest in rootfs.storage under the KATA_VIRTUAL_VOLUME_IMAGE_GUEST_PULL key (which maps to image_guest_pull?) as a json object and the image name in storage.source?
yes, absolutely correct.

**Comment by stevenhorsman** [comment]
> Hi, @stevenhorsman . I have updated my commits. The first eight commits are same as https://github.com/kata-containers/kata-containers/pull/7688. The last three commits implement the feature of pulling image in the guest. Please help me to review again.

Does mean that the best approach is to get #7688 merged first then rebase this? If so I can prioritising review that one - sorry if I've been looking at them in the wrong order.

**Comment by ChengyuZhu6** [comment]
> > Hi, @stevenhorsman . I have updated my commits. The first eight commits are same as #7688. The last three commits implement the feature of pulling image in the guest. Please help me to review again.
> 
> Does mean that the best approach is to get #7688 merged first then rebase this? If so I can prioritising review that one - sorry if I've been looking at them in the wrong order.

Yes, it does make sense to prioritize getting #7688 merged first. Once that's done, we can rebase this one more easily.  I appreciate your understanding.

**Comment by fidencio** [comment]
@ChengyuZhu6, I see nothing that jumps to my eyes as wrong on this PR, good work!

However, I'm dense, I really am, and I would like to see it working as expected before we merged it.
As the [nydus-snapshotter dependency](https://github.com/containerd/nydus-snapshotter/pull/499) has been solved, do you think we can have a unit test tied to this PR, as part of the tests repo, so we can exercise the code you're introduced?

**Comment by ChengyuZhu6** [comment]
> do you think we can have a unit test tied to this PR, as part of the tests repo, so we can exercise the code you're introduced?

I'm preparing the unit tests with this PR and https://github.com/kata-containers/kata-containers/pull/7688 in `kata-containers/tests` repo.

**Comment by fidencio** [comment]
> I'm preparing the unit tests with this PR and #7688 in `kata-containers/tests` repo.

I think we need kubernetes integration tests for both cases.

**Comment by ChengyuZhu6** [comment]
> > I'm preparing the unit tests with this PR and #7688 in `kata-containers/tests` repo.
> 
> I think we need kubernetes integration tests for both cases.

Yes, that's exactly what I wanted to say.

**Comment by ChengyuZhu6** [comment]
> Let's wait till TDX tests pass, and we should be good to merge the 2 PRs coming from you.



> Let's wait till TDX tests pass, and we should be good to merge the 2 PRs coming from you.



> Let's wait till TDX tests pass, and we should be good to merge the 2 PRs coming from you.

By the way, after https://github.com/kata-containers/kata-containers/pull/7688 merged, I need to rebase this PR.

**Comment by ChengyuZhu6** [comment]
/test-ubuntu-20.04-x86_64-cc_cri_containerd_k8s

File: src/runtime/virtcontainers/container.go
Comment (jiangliu) [inline src/runtime/virtcontainers/container.go:824]: The `else if` branch can be removed.

File: src/runtime/virtcontainers/kata_agent.go
Comment (jiangliu) [inline src/runtime/virtcontainers/kata_agent.go:1575]: How about simplifying it as `if virtualVolumeInfo.Source == ""`?

File: src/runtime/virtcontainers/kata_agent.go
Comment (jiangliu) [inline src/runtime/virtcontainers/kata_agent.go:1598]: Field `Driver` is set twice, so please remove line 1598.

File: src/runtime/virtcontainers/kata_agent.go
Comment (ChengyuZhu6) [reply src/runtime/virtcontainers/kata_agent.go]: Fixed.

File: src/runtime/virtcontainers/kata_agent.go
Comment (ChengyuZhu6) [reply src/runtime/virtcontainers/kata_agent.go]: Fixed.

File: src/runtime/virtcontainers/container.go
Comment (ChengyuZhu6) [reply src/runtime/virtcontainers/container.go]: Fixed.

**Comment by ChengyuZhu6** [comment]
tdx qemu test passed: http://10.112.240.228:8080/job/kata-containers-CCv0-qemu-TDX-CRI_CONTAINERD_K8S-PR/1249/

File: src/agent/src/image_rpc.rs
Comment (jiangliu) [inline src/agent/src/image_rpc.rs:143]: How about `get_security_config()`?

File: src/agent/src/image_rpc.rs
Comment (jiangliu) [inline src/agent/src/image_rpc.rs:170]: It's more effect to take `Option<&str>` than `&Option<&str>`.

File: src/agent/src/image_rpc.rs
Comment (jiangliu) [inline src/agent/src/image_rpc.rs:177]: Seems we need to refine image-rs for `pull_image()` to take `Option<&str>`.

Comment (jiangliu) [reply src/agent/src/image_rpc.rs:177]: But it's not blocking issue so we can refine it later.

File: src/agent/src/image_rpc.rs
Comment (jiangliu) [inline src/agent/src/image_rpc.rs:202]: How about `pull_image_for_container()`?

File: src/agent/src/image_rpc.rs
Comment (jiangliu) [inline src/agent/src/image_rpc.rs:204]: It's unsafe to detect pause container to check `pause` prefix in image name.
On the other handle, CRI provides an annotation about container type: "pause" vs "normal". So we should make use of the annotation for safety.
Another TODO:)

File: src/agent/src/image_rpc.rs
Comment (jiangliu) [inline src/agent/src/image_rpc.rs:205]: Why do we add an additional level in the directory path, such as `pause` and `images`?

File: src/agent/src/storage/image_pull_handler.rs
Comment (jiangliu) [inline src/agent/src/storage/image_pull_handler.rs:23]: It would be better to return `ImagePullVolume` instead of `ImagePullVolume.metadata`. We may make use of other fields later.

File: src/agent/src/storage/image_pull_handler.rs
Comment (jiangliu) [inline src/agent/src/storage/image_pull_handler.rs:56]: We should pass `image_metadata` to `pull_image_in_create_container()`.

File: src/agent/src/image_rpc.rs
Comment (ChengyuZhu6) [reply src/agent/src/image_rpc.rs]: Updated.

File: src/agent/src/image_rpc.rs
Comment (ChengyuZhu6) [reply src/agent/src/image_rpc.rs]: Fixed.

File: src/agent/src/image_rpc.rs
Comment (ChengyuZhu6) [reply src/agent/src/image_rpc.rs]: Sure.

File: src/agent/src/image_rpc.rs
Comment (ChengyuZhu6) [reply src/agent/src/image_rpc.rs]: Updated.

File: src/agent/src/image_rpc.rs
Comment (ChengyuZhu6) [reply src/agent/src/image_rpc.rs:204]: Sure.

File: src/agent/src/storage/image_pull_handler.rs
Comment (jiangliu) [inline src/agent/src/storage/image_pull_handler.rs:60]: How about adding `ro` flag?

File: src/agent/src/image_rpc.rs
Comment (jiangliu) [reply src/agent/src/image_rpc.rs:204]: We should detect container type by checking `io.kubernetes.cri.container-type` in ImagePullVolume.Metadata.

Comment (ChengyuZhu6) [reply src/agent/src/image_rpc.rs]: I have already handled it in runtime: 
```
if container_type == "sandbox" {
			image_ref = "pause"
		} else {
			image_ref = container_annotations["io.kubernetes.cri.image-name"]
			if image_ref == "" {
				return nil, fmt.Errorf("Failed to get image name from annotations")
			}
		}
```
I believe it's unnecessary to inform the Image service whether it's dealing with a container or a sandbox.

File: src/agent/src/storage/image_pull_handler.rs
Comment (ChengyuZhu6) [reply src/agent/src/storage/image_pull_handler.rs]: Fixed.

File: src/agent/src/storage/image_pull_handler.rs
Comment (ChengyuZhu6) [reply src/agent/src/storage/image_pull_handler.rs]: Fixed.

File: src/agent/src/storage/image_pull_handler.rs
Comment (ChengyuZhu6) [reply src/agent/src/storage/image_pull_handler.rs]: Fixed.

File: src/agent/src/image_rpc.rs
Comment (jiangliu) [reply src/agent/src/image_rpc.rs]: I mean we need a reliable way to detect pause image instead of checking the "pause" prefix in image ref.
Checking "io.kubernetes.cri.container-type" is a reliable way to detect pause image.

Comment (jiangliu) [reply src/agent/src/image_rpc.rs]: `image.contains("pause")` is really unreliable.

Comment (ChengyuZhu6) [reply src/agent/src/image_rpc.rs]: Sure.

Comment (ChengyuZhu6) [reply src/agent/src/image_rpc.rs]: Done.

**Review by jiangliu — APPROVED** [review]

File: src/agent/src/image_rpc.rs
Comment (stevenhorsman) [inline src/agent/src/image_rpc.rs:201]: nit: Should `bunld` be `bundle` here?

Comment (ChengyuZhu6) [reply src/agent/src/image_rpc.rs]: Fixed.

**Review by stevenhorsman — APPROVED** [review]
The code looks good to me, apart from a potential issue with the `main` nydus tests and the ci tests in https://github.com/kata-containers/tests/pull/5764 are working for x86 and s390x
```
ok 4 [cc][agent][kubernetes][containerd] Test can pull an image inside the guest with remote-snapshotter
```

**Comment by ChengyuZhu6** [comment]
I have created a PR https://github.com/kata-containers/kata-containers/pull/7961 to bump nydus and nydus-snapshotter. So I remove the related commits in this PR.

**Comment by fidencio** [comment]
Neither SEV nor TDX tests started.

I'm going ahead and merging this one as all the other tests have passed, and this will unblock further tests on the peer-pods side, and further development of the Kata Containers tests for this feature.

**Pull request merged**
